keps/sig-network/2079-network-policy-port-range/README.md

```diff
@@ -0,0 +1,434 @@
+  - [Monitoring Requirements](#monitoring-requirements)
+  - [Dependencies](#dependencies)
+  - [Scalability](#scalability)
+  - [Troubleshooting](#troubleshooting)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+
+## Summary
+
+Today the ``ports``field in ingress and egress network policies is an array 
+that needs a declaration of each single port to be contemplated. This KEP 
+proposes to add a new field that allows a declaration of a port range, 
+simplifying the creation of rules with multiple ports.
+
+## Motivation
+
+NetworkPolicy object is a complex object, that allows a developer to specify 
+what's the traffic behavior expected of the application and allow/deny 
+undesired traffic.
+
+There are a number of user issues like [kubernetes #67526](https://github.com/kubernetes/kubernetes/issues/67526) 
+and [kubernetes #93111](https://github.com/kubernetes/kubernetes/issues/93111) 
+where users expose the need to create a policy that allow a range of ports but some 
+specific port, or also cases that a user wants to create a policy that allows 
+the egress to other cluster to the NodePort range (eg 32000-32768) and in this case, 
+the rule should be created specifying each port separately, as:
+
+```
+spec:
+  egress:
+  - ports:
+    - protocol: TCP
+      port: 32000
+    - protocol: TCP
+      port: 32001
+    - protocol: TCP
+      port: 32002
+    - protocol: TCP
+      port: 32003
+[...]
+    - protocol: TCP
+      port: 32768
+```
+
+So for the user:
+* To allow a range of ports, each of them must be declared as an item from 
+``ports`` array
+* To make an exception needs a declaration of all ports but the exception
+
+Adding a new ``PortRange`` field inside the ``ports`` will allow a simpler 
+creation of NetworkPolicy to the user.
+
+### Goals
+
+Add Port Range field in a NetworkPolicy
+
+### Non-Goals
+
+N/A
+
+## Proposal
+
+In NetworkPolicy specification, inside ``NetworkPolicyIngressRule`` and 
```

Comment:
If I read correctly this is not a compatible change.

If the user specifies `portRanges` and does not specify `ports` they hit this part of the spec:

```
type NetworkPolicyIngressRule struct {
    // List of ports which should be made accessible on the pods selected for this
    // rule. Each item in this list is combined using a logical OR. If this field is
    // empty or missing, this rule matches all ports (traffic not restricted by port).
    // If this field is present and contains at least one item, then this rule allows
    // traffic only if the traffic matches at least one port in the list.
    // +optional
    Ports []NetworkPolicyPort
```

Specifically "If this field is empty or missing, this rule matches all ports"

In other words, a back-rev implementation which does not understand the new field will fail-open, which is a no-no.

What if, instead, you added a "size" or "rangeSize" or "endPort" field to `NetworkPolicyPort` which converted the single port into a range.  That way, a back-rev implementation who did not know about the new field would allow access to/from the single "base" port"  and nothing else.  Still not the correct interpretation, but it fails closed.

I *think* that could work.  Could use extra brains on it.

Reply:
Same problem from namespaceName selector, right? What if the CNI does not know the new field, got it.

I see value here in having the endPort (don't think rangeSize is very intuitive, as you need to make some 'math' to calculate big ranges)..

Reply:
Calico's datamodel supports port ranges by specifying the start and end port; this has worked well and I can't remember anyone having any issues grokking it; @thockin's idea sounds like a least-bad way to handle back compat.

Agree with @rikatz that rangeSize is not intuitive.

Could the struct be renamed to `NetworkPolicyPortRange` and still preserve back-compat (since the struct name doesn't actually feature in the JSON/YAML).

Reply:
> Calico's datamodel supports port ranges by specifying the start and end port; this has worked well and I can't remember anyone having any issues grokking it; @thockin's idea sounds like a least-bad way to handle back compat.
> 
> Agree with @rikatz that rangeSize is not intuitive.
> 
> Could the struct be renamed to `NetworkPolicyPortRange` and still preserve back-compat (since the struct name doesn't actually feature in the JSON/YAML).

renaming the struct would probably break anyone referring to the type downstream?

Reply:
Yes, I guess. I'm afraid of someone referencing the struct name and breaking things. I think probably the best now is to specify the endPort (assuming the endPort is also part of the range)

Reply:
> renaming the struct would probably break anyone referring to the type downstream?

Could leave a type alias in its place.  No big deal either way.